File: c++/3.longest-substring-without-repeating-characters.cpp
/*
 * @Author: AndyWong 
 * @Date: 2018-09-19 16:56:48 
 * @Last Modified by:   AndyWong 
 * @Last Modified time: 2018-09-19 16:56:48 
 */
/*
 * [3] Longest Substring Without Repeating Characters
 *
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
 *
 * algorithms
 * Medium (25.02%)
 * Total Accepted:    578.4K
 * Total Submissions: 2.3M
 * Testcase Example:  '"abcabcbb"'
 *
 * Given a string, find the length of the longest substring without repeating
 * characters.
 *
 *
 * Example 1:
 *
 *
 * Input: "abcabcbb"
 * Output: 3
 * Explanation: The answer is "abc", with the length of 3.
 *
 *
 *
 * Example 2:
 *
 *
 * Input: "bbbbb"
 * Output: 1
 * Explanation: The answer is "b", with the length of 1.
 *
 *
 *
 * Example 3:
 *
 *
 * Input: "pwwkew"
 * Output: 3
 * Explanation: The answer is "wke", with the length of 3.
 * ⁠            Note that the answer must be a substring, "pwke" is a
 * subsequence and not a substring.
 *
 *
 *
 *
 *
 */

#include <iostream>
#include <string>
#include <vector>
using namespace std;
class Solution
{
  public:
    int findSame(string temp)
    {
        int length = temp.length();
        for (int i = 0; i < length; i++)
            for (int j = i + 1; j < length; j++)
                if (temp[i] == temp[j])
                {
                    return j;
                }
        return -1;
    }
    int lengthOfLongestSubstring1(string s)
    {
        int i = 0, j = s.length(), length = s.length(), count = 0;
        for (i = 0; i < length; i++)
        {
            for (j = s.length() - i; j >= count; j--)
            {
                string temp = s.substr(i, j);
                int tempLength = temp.length();
                int isSame = findSame(temp);
                if ((isSame == -1) && (count < tempLength))
                    count = tempLength;
                else
                    j = isSame + 1;
            }
        }
        return count;
    }
    int lengthOfLongestSubstring(string s)
    {
        vector<int> mymap(255, -1);
        int len = 0, i = -1, tmp;
        int length = s.length();
        for (int j = 0; j < length; j++)
        {
            tmp = mymap[s.at(j)];
            i = std::max(tmp, i);
            mymap[s.at(j)] = j;
            len = std::max(len, (j - i));
        }
        return len;
    }
};

File: python/3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#
# https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
#
# algorithms
# Medium (28.74%)
# Likes:    8084
# Dislikes: 494
# Total Accepted:    1.4M
# Total Submissions: 4.7M
# Testcase Example:  '"abcabcbb"'
#
# Given a string, find the length of the longest substring without repeating
# characters.
#
#
# Example 1:
#
#
# Input: "abcabcbb"
# Output: 3
# Explanation: The answer is "abc", with the length of 3.
#
#
#
# Example 2:
#
#
# Input: "bbbbb"
# Output: 1
# Explanation: The answer is "b", with the length of 1.
#
#
#
# Example 3:
#
#
# Input: "pwwkew"
# Output: 3
# Explanation: The answer is "wke", with the length of 3.
# ⁠            Note that the answer must be a substring, "pwke" is a
# subsequence and not a substring.
#
#
#
#
#
# 抓好前后两个指针度量的位置，灵活使用hash表来统计现状


# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        result, i, j = 0, 0, 0
        resultDict = {}
        for c in s:
            j += 1
            resultDict[c] = resultDict.get(c, 0) + 1
            while resultDict[c] > 1:
                resultDict[s[i]] -= 1
                i += 1
            result = max(result, j - i)
        return result

        # @lc code=end
